### PR TITLE
Feature/fix recursion

### PR DESCRIPTION
**Pull request by gianfurrer**

### CODE REVIEW

File: INCode/diagramconfiguration.py [hunk 1]
```diff
@@ -5,59 +5,82 @@
...
+            parent.referenced_items_.append(self)
+        self.index_ = callable.index_
         self.callable_id_ = callable.get_id()
+        self.state_ = Qt.Unchecked
+        self.sender_ = callable._get_sender(callable.cursor_)
+        self.referenced_items_ = []
         self.setText(TreeColumns.FIRST_COLUMN, callable.get_name())
         self.setFlags(self.flags() | Qt.ItemIsUserCheckable)
-        self.setCheckState(TreeColumns.FIRST_COLUMN, CallableTreeItem.get_check_state_(callable))
+        self.setCheckState(TreeColumns.FIRST_COLUMN, self.get_check_state())
```
Comment (rknuus) [inline INCode/diagramconfiguration.py]: On line 21 the state is set to `Qt.Unchecked`. Not sure whether call to `self.setCheckState` is still necessary:

```suggestion
```

File: INCode/diagramconfiguration.py [hunk 2]
```diff
@@ -73,16 +96,11 @@ def revealChildren(self):
...
         callable = current_item.get_callable()
-        if not callable.is_definition():
-            callable = self.index_.load_definition(callable)
-            if not callable.is_definition():
-                return  # TODO(KNR): really?
 
         for child in callable.get_referenced_callables():
             child.initialize()  # lazy load the referenced callables
-            CallableTreeItem(child, self.index_, current_item)
+            child_tree_item = CallableTreeItem(child, current_item)
```
Comment (rknuus) [inline INCode/diagramconfiguration.py:97]: `child_tree_item` is not used. Suggest:

```suggestion
            CallableTreeItem(child, current_item)
```

File: INCode/models.py [hunk 3]
```diff
@@ -64,8 +64,6 @@ def load_definition(self, declaration):
         candidates = Index._gen_search(declaration.get_name(), translation_units)
         for candidate in candidates:
             self.load(candidate)
-        if declaration.is_included():
```
Comment (rknuus) [inline INCode/models.py:67]: This special case was necessary when loading a definition over an already included declaration. How is this case now handled in `CallableTreeItem`?

File: tests/diagramconfiguration_test.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    diagram = export_callable_tree_item.export()
+    expected_diagram = '''@startuml
+
+B -> B: void m()
+
+@enduml'''
+
+    assert diagram == expected_diagram
+
+def setup_diagram_configuration(directory):
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py]:
```suggestion
def build_diagram_configuration(directory):
```

**Review by rknuus — COMMENTED** [review]
Please check my suggestions and whether line 26 of `CallableTreeItem` is still necessary. Also, let's peer-review the tests. Cheers.

File: INCode/diagramconfiguration.py (hunk 1 above)
Comment (gianfurrer) [reply INCode/diagramconfiguration.py]: Without this line there is no checkbox in the GUI..
Although I could remove `self.state_` completely and replace it with `self.checkState(TreeColumns.FIRST_COLUMN)` and `self.setCheckState(TreeColumns.FIRST_COLUMN, Qt.Checked/Qt.Unchecked)`

File: INCode/models.py (hunk 3 above)
Comment (gianfurrer) [reply INCode/models.py:67]: The `CallableTreeItem` contains the include information and doesn't get overridden when loading a definition over a declaration.

File: tests/diagramconfiguration_test.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
-from INCode.models import CompilationDatabases, Index, Callable
+from INCode.models import Callable
 from INCode.diagramconfiguration import DiagramConfiguration, CallableTreeItem
-from test_environment_generation import *
+from tests.test_environment_generation import *
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py]: Wild-card imports are not recommended by [PEP 8](https://www.python.org/dev/peps/pep-0008/#imports).

File: tests/diagramconfiguration_test.py [hunk 6]
```diff
@@ -179,7 +178,9 @@ class B {
     }
 
 private:
-    void p() {}
+    void p() {
+        p()
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py]: Terminate line by `;`

File: tests/diagramconfiguration_test.py [hunk 7]
```diff
@@ -221,14 +278,36 @@ class B {
...
 
     return DiagramConfiguration(entry_point)
 
-def test__reveal_children__has_children_after_reveal(directory):
+
+def test__diagram_configuration__reveal_children_of_item_without_references__has_no_children_after_reveal(directory):
+    diagram_configuration = setup_diagram_configuration(directory, '''
+class B {
+public:
+    void a() {
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py]: Suggest to remove method `b` and call to `b` in `a`.

File: tests/diagramconfiguration_test.py [hunk 8]
```diff
@@ -221,14 +278,36 @@ class B {
...
+''')
+
+    child_tree_item = diagram_configuration.entry_point_item_.referenced_items_[0]
+    diagram_configuration.tree_.setCurrentItem(child_tree_item)
+
+    diagram_configuration.revealChildren()
+    assert child_tree_item.childCount() == 0
+
+
+def test__diagram_configuration__reveal_children__has_children_after_reveal(directory):
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py:331]: `revealChildren` not called

File: tests/diagramconfiguration_test.py [hunk 9]
```diff
@@ -238,7 +317,7 @@ def test__reveal_children__has_children_after_reveal(directory):
     assert child_tree_item.childCount() > 0
 
 
-def test__reveal_children__no_duplicates_after_multiple_reveal(directory):
+def test__diagram_configuration__reveal_children__no_duplicates_after_multiple_reveal(directory):
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py:341]: `revealChildren` not called multiple times

File: tests/diagramconfiguration_test.py [hunk 10]
```diff
@@ -250,24 +329,38 @@ def test__reveal_children__no_duplicates_after_multiple_reveal(directory):
...
+    entry_point_item = diagram_configuration.entry_point_item_
+    diagram_configuration.tree_.setCurrentItem(entry_point_item)
+
+    with patch.object(entry_point_item, 'get_callable') as mock:
+        diagram_configuration.revealChildren()
+
+    mock.assert_not_called()
+
+
+def test__diagram_configuration__reveal_if_no_item_selected__no_action(directory):
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py:354]: Consider to use the same test approach for tests `test__diagram_configuration__reveal_already_revealed_item__no_action` and `test__diagram_configuration__reveal_if_no_item_selected__no_action` as both are supposed to trigger no action

File: tests/model_test.py [hunk 11]
```diff
@@ -3,7 +3,7 @@
 from clang.cindex import CursorKind
 from INCode.models import Callable, CompilationDatabases, File, Index
 from unittest.mock import MagicMock
-from test_environment_generation import *
+from tests.test_environment_generation import *
```
Comment (rknuus) [inline tests/model_test.py]: PEP 8

File: INCode/diagramconfiguration.py [hunk 12]
```diff
@@ -12,52 +13,79 @@ class TreeColumns(IntEnum):
...
+            child_callable = child_item.get_callable()
+            sender = self.get_sender() if self.is_included() else parent_sender
+            if child_item.is_included():
+                diagram += '{} -> {}: {}\n'.format(sender,
+                                                   child_item.get_sender(),
+                                                   child_callable.get_name())
+            diagram += child_item.export_relations_(sender)
+        return diagram
+
+    def get_sender(self):
```
Comment (rknuus) [inline INCode/diagramconfiguration.py]: Please merge into function case of `Callable._get_sender`

File: INCode/diagramconfiguration.py [hunk 13]
```diff
@@ -12,52 +13,79 @@ class TreeColumns(IntEnum):
...
 class CallableTreeItem(QTreeWidgetItem):
-    def __init__(self, callable, index, parent=None):
+    def __init__(self, callable, parent=None):
         super(QTreeWidgetItem, self).__init__(parent)
-        self.index_ = index
+
+        self.common_path_ = ''
+        if isinstance(parent, CallableTreeItem):
+            parent.referenced_items_.append(self)
+            self.common_path_ = parent.common_path_
```
Comment (rknuus) [inline INCode/diagramconfiguration.py]: Consider to factor the common path out to a central location. Related to https://github.com/rknuus/INCode/pull/1/files#r251864190

File: INCode/models.py [hunk 14]
```diff
@@ -61,11 +61,9 @@ def load(self, file):
 
     def load_definition(self, declaration):
         translation_units = Index._gen_open(self.compilation_databases_.get_files())
-        candidates = Index._gen_search(declaration.get_name(), translation_units)
+        candidates = Index._gen_search(declaration.cursor_.spelling, translation_units)
```
Comment (rknuus) [inline INCode/models.py:69]: Please add tests:
- with identical function signatures spread over multiple lines in the H and CPP file
- with entire function signature on same line in H and function signature spread over multiple lines in CPP file

File: tests/diagramconfiguration_test.py [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+    expected_diagram = '''@startuml
+
+foo.cpp -> baz.cpp: void baz()
+
+@enduml'''
+
+    assert diagram == expected_diagram
+
+
+def test__callable_tree_item__export_definition_loaded_over_declaration__export_correct_diagram(two_translation_units):
```
Comment (rknuus) [inline tests/diagramconfiguration_test.py]: Please provide a similar test for methods

File: INCode/diagramconfiguration.py (hunk 2 above)
Comment (gianfurrer) [reply INCode/diagramconfiguration.py:97]: object is required below due to a recent change

File: INCode/entrydialog.py [hunk 16]
```diff
@@ -33,6 +33,7 @@ def __init__(self, parent=None):
         self.entry_file_list_.setModel(self.entry_files_)
         self.entry_file_selection_ = self.entry_file_list_.selectionModel()
         self.entry_file_selection_.currentChanged.connect(self.onSelectEntryFile)
+        self.common_path = ''
```
Comment (rknuus) [inline INCode/entrydialog.py]: Got moved to Index, no longer required

File: INCode/models.py [hunk 17]
```diff
@@ -196,19 +185,21 @@ def _is_a_call(cursor):
...
-        elif cursor.kind == CursorKind.CXX_METHOD:
-            return self._get_class(cursor)
-        return '{} is not supported'.format(cursor.kind)
+    def _get_class(self):
+        if self.cursor_.kind == CursorKind.CXX_METHOD:
+            return self.cursor_.semantic_parent.displayname
+        return '{} is not supported'.format(self.cursor_.kind)
+
+    def _get_name(self, get_function_signature=False):
+        if self.cursor_.kind == CursorKind.FUNCTION_DECL or get_function_signature:
```
Comment (rknuus) [inline INCode/models.py]: What's the purpose of `or get_function_signature`?

File: INCode/models.py (hunk 14 above)
Comment (gianfurrer) [reply INCode/models.py:69]: done

File: INCode/models.py (hunk 17 above)
Comment (gianfurrer) [reply INCode/models.py]: Will rename to `name_only`.

File: tests/diagramconfiguration_test.py (hunk 10 above)
Comment (gianfurrer) [reply tests/diagramconfiguration_test.py:354]: done

File: tests/diagramconfiguration_test.py (hunk 8 above)
Comment (rknuus) [reply tests/diagramconfiguration_test.py:331]: ignore, my bad

File: tests/diagramconfiguration_test.py (hunk 9 above)
Comment (rknuus) [reply tests/diagramconfiguration_test.py:341]: ignore, my bad

File: tests/diagramconfiguration_test.py (hunk 4 above)
Comment (rknuus) [reply tests/diagramconfiguration_test.py]: minor comment, reject

**Pull request merged**